### DISCUSSION TITLE
What's the best way to customise envoy bootstrap template globally?

### DISCUSSION THREAD

**Discussion by qudongfang** [istio/istio#52434] [Q&A]
We want to enable the [enable_deferred_creation_stats](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/bootstrap/v3/bootstrap.proto#envoy-v3-api-msg-config-bootstrap-v3-bootstrap-deferredstatoptions) option to reduce the metrics volume as discussed in https://github.com/istio/istio/issues/49868.

`enable_deferred_cluster_creation ` was supported by https://github.com/istio/istio/pull/50785, but not `enable_deferred_creation_stats`

We can enable  `enable_deferred_creation_stats` with the [custom-bootstrap](https://github.com/istio/istio/blob/master/samples/custom-bootstrap/README.md) per workload.

It would be great if there is an easier way to enable the option globally.

We considered envoyfilter with BOOTSTRAP, but it is DEPRECATED.

Any input would be greatly appreciated!

**Comment by ramaraochavali** [comment]
have you tried this and found to be helpful? We are looking for some data to make it default.

**Comment by ramasai1** [comment]
Hi, yes we've tried it and have found it to be helpful (I work with @qudongfang). Some stats about the cluster we've tried it on:
```
$ istioctl pc clusters <pod> | awk '{print $1}' | wc -l
      85 (adjusted for the first row being the name)

$ istioctl pc route <pod> | awk '{print $2}' | wc -l
      84

$ istioctl pc listeners <pod> | awk '{print $2}' | uniq | wc -l
      15
```

`istioctl` version:
```
$ istioctl version
client version: 1.22.3
istiod version: 1.22.2
data plane version: 1.22.2 (60 proxies)
```

Metrics stats without `enable_deferred_creation_stats` from this pod:
```
$ istioctl pc bootstrap <pod> | grep -i deferred

$ istioctl x envoy-stats <pod> | wc -l
    6130
```

Metrics stats with `enable_deferred_creation_stats` from this pod:
```
$ istioctl pc bootstrap <pod> | grep -i deferred
        "deferredStatOptions": {
            "enableDeferredCreationStats": true

$ istioctl x envoy-stats <pod> | wc -l
    3692
```

In our opinion it'd be beneficial if we made this the default (or offered an option to toggle it via an env var similar to `enable_deferred_cluster_creation`.

**Comment by ramasai1** [comment]
Hi @ramaraochavali just checking in to see if you had a chance to look at my previous response, thanks.

**Comment by ramaraochavali** [comment]
Sorry for the delay. Can you please add any memory/cpu stats with and without this?

**Comment by ramasai1** [comment]
Hello @ramaraochavali, here is a memory consumption chart from a busy deployment on a busy cluster:
<img width="1634" alt="image" src="https://github.com/user-attachments/assets/3369bc0a-2d0a-4e3f-9cb4-5d83b5695fd5">

The restarts are after I applied the custom bootstrap. That looks like a solid improvement. Here's a graph showing CPU usage (calculated using the `container_cpu_usage_seconds_total` metric):
<img width="1682" alt="image" src="https://github.com/user-attachments/assets/cd35ba83-cd04-4322-8ac4-103c49120342">

The timing should be the same as the memory chart. Please let me know if you need anything else.

**Comment by ramaraochavali** [comment]
Thank you that makes sense. I will push a PR to enable it

**Comment by ramasai1** [comment]
@ramaraochavali To add some more data, this is a steady state view of memory (the earlier image may have been a bit misleading due to grabbing the screenshot early):

<img width="1270" alt="image" src="https://github.com/user-attachments/assets/e74d863b-2ca5-438a-9767-6dff77076664">

Time should still be the same - around 02:50 is when the new bootstrap went into effect. The steady state memory usage is clocking in at about 300-320 MB whereas the before the change it was around 360-380 MB. Still a decent decrease.
